synapse/replication/tcp/handler.py

```diff
@@ -202,33 +203,32 @@ async def _add_command_to_stream_queue(
             logger.error("Got %s for unknown stream: %s", cmd.NAME, stream_name)
             return
 
-        # if we're already processing this stream, stick the new command in the
-        # queue, and we're done.
+        queue.append((cmd, conn))
+
+        # if we're already processing this stream, there's nothing more to do:
+        # the new entry on the queue will get picked up in due course
         if stream_name in self._processing_streams:
-            queue.append((cmd, conn))
             return
 
-        # otherwise, process the new command.
+        # fire off a background process to start processing the queue.
+        run_as_background_process(
+            "process-replication-data", self._unsafe_process_queue, conn, cmd
+        )
 
-        # arguably we should start off a new background process here, but nothing
-        # will be too upset if we don't return for ages, so let's save the overhead
-        # and use the existing logcontext.
+    async def _unsafe_process_queue(self, stream_name: str):
+        """Processes the command queue for the given stream, until it is empty
 
+        Does not check if there is already a thread processing the queue, hence "unsafe"
+        """
         self._processing_streams.add(stream_name)
```

Comment:
Can we add another guard here to test `stream_name in self._processing_streams` in case there is a race? I don't think there should be but it requires understanding how `run_as_background_process` works

Reply:
it's a pattern we use pretty widely across the codebase. I've added an assertion.